drivers/sensor/bmm150/bmm150.c

```diff
@@ -0,0 +1,588 @@
+			      BMM150_REG_REP_Z, &reg_val) < 0) {
+		return -EIO;
+	}
+
+	data->rep_z = BMM150_REGVAL_TO_REPZ((int)(reg_val));
+
+	return 0;
+}
+
+static int bmm150_compute_max_odr(struct device *dev, int rep_xy,
+				  int rep_z, int *max_odr)
+{
+	struct bmm150_data *data = dev->driver_data;
+
+	if (rep_xy == 0) {
+		if (data->rep_xy <= 0) {
+			if (bmm150_read_rep_xy(dev) < 0) {
+				return -EIO;
+			}
+		}
+		rep_xy = data->rep_xy;
+	}
+
+	if (rep_z == 0) {
+		if (data->rep_z <= 0) {
+			if (bmm150_read_rep_z(dev) < 0) {
+				return -EIO;
+			}
+		}
+		rep_z = data->rep_z;
+	}
+
+	*max_odr = 1000000 / (145 * rep_xy + 500 * rep_z + 980);
+
+	return 0;
+}
+#endif
+
+#if defined(BMM150_SET_ATTR_REP)
+static int bmm150_read_odr(struct device *dev)
+{
+	struct bmm150_data *data = dev->driver_data;
+	const struct bmm150_config *config = dev->config->config_info;
+	u8_t i, odr_val, reg_val;
+
+	if (i2c_reg_read_byte(data->i2c, config->i2c_slave_addr,
+			      BMM150_REG_OPMODE_ODR, &reg_val) < 0) {
+		return -EIO;
+	}
+
+	odr_val = (reg_val & BMM150_MASK_ODR) >> BMM150_SHIFT_ODR;
+
+	for (i = 0; i < ARRAY_SIZE(bmm150_samp_freq_table); ++i) {
+		if (bmm150_samp_freq_table[i].reg_val == odr_val) {
+			data->odr = bmm150_samp_freq_table[i].freq;
+			return 0;
+		}
+	}
+
+	return -ENOTSUP;
+}
+#endif
+
+#if defined(CONFIG_BMM150_SAMPLING_REP_XY)
+static int bmm150_write_rep_xy(struct device *dev, int val)
+{
+	struct bmm150_data *data = dev->driver_data;
+	const struct bmm150_config *config = dev->config->config_info;
+
+	if (i2c_reg_update_byte(data->i2c, config->i2c_slave_addr,
+				BMM150_REG_REP_XY,
+				BMM150_REG_REP_DATAMASK,
+				BMM150_REPXY_TO_REGVAL(val)) < 0) {
+		return -EIO;
+	}
+
+	data->rep_xy = val;
+
+	return 0;
+}
+#endif
+
+#if defined(CONFIG_BMM150_SAMPLING_REP_Z)
+static int bmm150_write_rep_z(struct device *dev, int val)
+{
+	struct bmm150_data *data = dev->driver_data;
+	const struct bmm150_config *config = dev->config->config_info;
+
+	if (i2c_reg_update_byte(data->i2c, config->i2c_slave_addr,
+				BMM150_REG_REP_Z,
+				BMM150_REG_REP_DATAMASK,
+				BMM150_REPZ_TO_REGVAL(val)) < 0) {
+		return -EIO;
+	}
+
+	data->rep_z = val;
+
+	return 0;
+}
+#endif
+
+/* Reference Datasheet 4.3.2 */
+static s32_t bmm150_compensate_xy(struct bmm150_trim_regs *tregs,
+				  s16_t xy, uint16_t rhall, bool is_x)
+{
+	s8_t txy1, txy2;
+	s16_t val;
+
+	if (xy == BMM150_XY_OVERFLOW_VAL) {
+		return INT32_MIN;
+	}
+
+	if (!rhall) {
+		rhall = tregs->xyz1;
+	}
+
+	if (is_x) {
+		txy1 = tregs->x1;
+		txy2 = tregs->x2;
+	} else {
+		txy1 = tregs->y1;
+		txy2 = tregs->y2;
+	}
+
+	val = ((s16_t)(((u16_t)((((s32_t)tregs->xyz1) << 14) / rhall)) -
+		       ((u16_t)0x4000)));
+	val = ((s16_t)((((s32_t)xy) * ((((((((s32_t)tregs->xy2) *
+		((((s32_t)val) * ((s32_t)val)) >> 7)) + (((s32_t)val) *
+		((s32_t)(((s16_t)tregs->xy1) << 7)))) >> 9) +
+		((s32_t)0x100000)) * ((s32_t)(((s16_t)txy2) +
+		((s16_t)0xA0)))) >> 12)) >> 13)) + (((s16_t)txy1) << 3);
```

Comment:
Please simplify these equations. There are too many casts, shifts, and magic numbers to be readable.